tuxemon/mod_manager.py

```diff
@@ -0,0 +1,190 @@
+                self.packages.append(package)
+
+        self.write_to_cache()
+
+    def list_packages(self):
+        """Returns package dictionary, either from the server or the cache"""
+        return self.packages
+
+    def download_package(self, name, release, repo=None, dont_extract=False, install_deps=True):
+        """Downloads the specified package"""
+        if repo is None:
+            repo = self.get_package_repo(name)
+
+            # Remove trailing slash
+            if repo[-1] == "/":
+                repo = repo[:-1]
+
+        url = str(repo) + f"/packages/{name}/releases/{str(release)}/download"
+        filename = os.path.join(paths.CACHE_DIR + f"/downloaded_packages/{name}.{release}.zip")
+
+        # Apperantly this function is ported from urllib from python2.
+        # Maybe replace this in the future?
+        # https://docs.python.org/3/library/urllib.request.html#urllib.request.urlretrieve
+        urllib.request.urlretrieve(url, filename=filename)
+
+        outfolder = os.path.join(paths.BASEDIR, "mods", f"{name}")
+
+        self.write_package_to_list(outfolder, name)
+
+        if not dont_extract:
+            self.extract(filename, outfolder)
+            with open(f"{outfolder}/meta.json", "w") as metafile:
+                meta = self.get_package_info(name, repo)
+                metafile.write(json.dumps(meta, indent=4, sort_keys=False))
+
+        if install_deps:
+            # This function calls download_package, might cause issues
+            self.install_dependencies(name, release, repo, dont_extract=dont_extract)
+
+    def install_dependencies(self, name, release, repo, dont_extract=False, symlink=True):
+        """
+        Same as the download_package(), but it includes dependency installing.
+        When symlink is True, dependency's files will be linked.
+        """
+        # Get info
+        meta = self.get_package_info(name, repo)
+        if "dependencies" in meta:
+            dependencies = meta["dependencies"]
+
+            for pack in dependencies:
+                self.download_package(pack, release, repo, dont_extract=dont_extract, install_deps=False)
+
+                # Symlink deps
+                mainfolder = os.path.join(paths.BASEDIR, "mods", name)
+                depfolder = os.path.join(paths.BASEDIR, "mods", pack)
+                symlink_missing(mainfolder, depfolder)
+        else: pass
+
+    def extract(self, file, outfolder):
+        """Extracts the specified zip archive to the mods directory"""
+        with zipfile.ZipFile(file) as zip_:
+            zip_.extractall(path=os.path.join(paths.BASEDIR, "mods", outfolder))
+
+    def get_package_info(self, name, repo):
+        """Get specified package info. Always downloads the info from the server."""
+        with urllib.request.urlopen(repo + f"/api/packages/{name}") as packages:
+            return json.loads(packages.read().decode("UTF-8"))
+
+    def get_package_repo(self, name):
```

Comment:
if we are allowing mods to be hosted by anyone, then we should assume all servers are malicious.  this means that the main try points to the server should be checking for broken or malicious values.  in the case of `name`, it is used to make file paths for writing files.  we should make sure that `name` cannot be used to escape from the tuxemon cache.